### ISSUE TITLE
Query regarding echo example

### ISSUE THREAD

**Issue by asimpatnaik** [libp2p/go-libp2p#368]
Hi,
Thanks a lot for the examples. I'm using the echo example for peer to peer data exchange. I have some basic queries:
1. When I run the example, it throws the warning - **WARNI net/identi: error reading identify message:  stream reset <autogenerated>:1**. Kindly let me know the cause as I haven't been able to figure that out.
2. I want to use the echo example for secure message exchange across domains. It will be great if you can suggest the changes so that I can achieve the same.

Thanks a lot.

Regards,
Asim.

**Comment by Stebalien** [comment]
> When I run the example, it throws the warning - WARNI net/identi: error reading identify message: stream reset :1. Kindly let me know the cause as I haven't been able to figure that out.

TL;DR: You can ignore this.

When two peers connect, they run a protocol we call "identify". We use this protocol to exchange information about what protocols each peer supports, the protocol versions, the other addresses at which the peer can be contacted, etc.

However, we don't *wait* for this identify exchange to complete before we allow other connections through. In the case of the echo example, the "echo" protocol is finishing *before* the identify protocol finishes. This means that the the echo client disconnects and exits before the identify protocol finishes so the echo server sees that the identify protocol has been aborted and logs a warning.

> I want to use the echo example for secure message exchange across domains. It will be great if you can suggest the changes so that I can achieve the same.

Could you expand on this?

**Comment by asimpatnaik** [comment]
Thank you for your response. It was indeed helpful.
Regarding my query on deploying and using the same across domains, let me try to explain my query with an example:
there r 3 nodes, say on a.com, b.com and c.com ... and each of them run an instance of echo server ... what code changes need to be done to the echo client so that we can echo msgs on the echo server instance?
Hope it helps you understand the query.

Regards,
Asim.

**Comment by Stebalien** [comment]
Do you want to broadcast to all three? Just one at a time? Pick one randomly?

You may want to import the `github.com/multiformats/go-multiaddr-dns` package on the client. This will allow it to connect to `a.com` by domain name instead of requiring an explicit IP address. That is, `/dns4/a.com/tcp/4001/ipfs/QmADotCom` instead of `/ip4/1.2.3.4/tcp/4001/ipfs/QmADotCom` (where `QmADotCom` is `a.com`'s peer ID).
